### DISCUSSION TITLE
Should `exclude_defaults=True` still include `Literal` fields with their default values?

### DISCUSSION THREAD

**Discussion by bjmc** [pydantic/pydantic#9108] [Feature Request]
I have a situation where I have several models that I want to to use as part of a [discriminated union](https://docs.pydantic.dev/latest/concepts/unions/#discriminated-unions-with-str-discriminators). These classes have a large number of optional fields, and I'd prefer not to clutter my serialized JSON ouput with `null`, `[]` and `{}`  values.

Unfortunately, I've found that if I call `model_dump_json(exclude_defaults=True)` the resulting output loses the discriminator field, and it wouldn't be easy to load/validate back to Pydantic models from that output. 

Here's a minimal reproduction of what I'm talking about...

```python
from enum import Enum
from typing import Literal, Optional
from pydantic import BaseModel, Field

class AnimalType(str, Enum):
    CAT = 'cat'
    DOG = 'dog'

class Pet(BaseModel):
    name: Optional[str] = None

class Cat(Pet):
    animal_type: Literal[AnimalType.CAT] = AnimalType.CAT

class Dog(Pet):
    animal_type: Literal[AnimalType.DOG] = AnimalType.DOG

class PetCarrier(BaseModel):
    contains: Cat | Dog = Field(discriminator='animal_type')

carrier = PetCarrier(contains=Cat(name='Cookie'))

print(carrier.model_dump_json(exclude_defaults=True))
# {"contains":{"name":"Cookie"}}
# Note that "animal_type" is absent from this JSON.
```

I'm marking this as a feature request, because I suppose I'm asking for a change in behavior, but I'd also be curious if anyone has a good workaround for this situation? 

Because a `Literal` field will, by definition, _always_ have its default value, I think a fair case can be made that `Literal`-typed fields should still be included in `model_dump()` output even with `exclude_defaults=True`. If somebody really wants to omit them, they could use the `exclude=` parameters to specially remove those fields.

**Comment by uriyyo** [comment]
Hi @bjmc,

You can define you custom serializer:
```py
from enum import Enum
from typing import Literal, Optional

from pydantic import BaseModel, Field, model_serializer


class AnimalType(str, Enum):
    CAT = "cat"
    DOG = "dog"


class Pet(BaseModel):
    name: Optional[str] = None

    @model_serializer(mode="wrap")
    def serialize(self, next_):
        dumped = next_(self)
        dumped["animal_type"] = self.animal_type

        return dumped


class Cat(Pet):
    animal_type: Literal[AnimalType.CAT] = AnimalType.CAT


class Dog(Pet):
    animal_type: Literal[AnimalType.DOG] = AnimalType.DOG


class PetCarrier(BaseModel):
    contains: Cat | Dog = Field(discriminator="animal_type")


carrier = PetCarrier(contains=Cat(name="Cookie"))
carrier.model_fields_set.add("animal_type")


print(carrier.model_dump_json(exclude_defaults=True))
```
```py
{"contains":{"name":"Cookie","animal_type":"cat"}}
```

**Reply by bjmc** [reply]
Thanks for the suggestion!

```python
@model_serializer(mode='wrap')
def include_literals(self, next_serializer):
    dumped = next_serializer(self)
    for name, field_info in self.model_fields.items():
        if typing.get_origin(field_info.annotation) == typing.Literal:
            dumped[name] = getattr(self, name)
    return dumped
```

I ended up with this, which seems to work. I do think the default behavior is unhelpful, but I can see the virtue of "doing what it says on the tin" (exclude defaults)
